src/Features/CSharp/Portable/NameLiteralArgument/CSharpNameLiteralArgumentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,128 @@
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.InvocationExpression,
+                SyntaxKind.ObjectCreationExpression, SyntaxKind.BaseConstructorInitializer,
+                SyntaxKind.ThisConstructorInitializer, SyntaxKind.ElementAccessExpression);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var syntaxTree = context.Node.SyntaxTree;
+            var parseOptions = (CSharpParseOptions)syntaxTree.Options;
+            if (parseOptions.LanguageVersion < LanguageVersion.CSharp7_2)
+            {
+                return;
+            }
+
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet is null)
+            {
+                return;
+            }
+            if (!optionSet.GetOption(CSharpCodeStyleOptions.PreferNamingLiteralArguments).Value)
+            {
+                return;
+            }
+
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(context.Node);
+            if (symbolInfo.Symbol is null)
+            {
+                return;
+            }
+
+            var parameters = symbolInfo.Symbol.GetParameters();
+            if (parameters.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+            var node = context.Node;
+            SeparatedSyntaxList<ArgumentSyntax> arguments;
+            switch (node)
+            {
+                case InvocationExpressionSyntax invocation:
+                    arguments = invocation.ArgumentList.Arguments;
+                    break;
+
+                case ObjectCreationExpressionSyntax creation:
+                    arguments = creation.ArgumentList.Arguments;
+                    break;
+
+                case ConstructorInitializerSyntax creation:
+                    arguments = creation.ArgumentList.Arguments;
+                    break;
+
+                case ElementAccessExpressionSyntax access:
+                    arguments = access.ArgumentList.Arguments;
+                    break;
+
+                default:
+                    return;
+            }
+
+            ReportDiagnosticIfNeeded(context, optionSet, parameters, arguments);
+        }
+
+        private void ReportDiagnosticIfNeeded(SyntaxNodeAnalysisContext context, OptionSet optionSet,
+            ImmutableArray<IParameterSymbol> parameters, SeparatedSyntaxList<ArgumentSyntax> arguments)
+        {
+            for (int i = 0; i < arguments.Count; i++)
+            {
+                var argument = arguments[i];
+                if (!IsPositionalArgument(argument))
+                {
+                    continue;
+                }
+
+                if (!argument.Expression.IsAnyLiteralExpression())
+                {
+                    continue;
+                }
+
+                if (parameters[i].IsParams)
```

Comment:
Slightly concerned about error cases (though that shouldn't probably happen if .Symbol was non-null).  But what if there are less parameters than arguments?

Reply:
Nit: combine all three checks into one.